Trace event support for Node.js

matthewloring opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines
##### Affected core subsystem(s)

<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

src
##### Description of change

<!-- Provide a description of the change below this comment. -->

Background: https://github.com/nodejs/diagnostics/issues/30, https://github.com/nodejs/diagnostics/issues/53

This PR adds support for trace-event tracing to Node.js. It provides a mechanism to centralize tracing information generated by V8, Node core, and userspace code. The PR contains a few components:
- A trace writer responsible for serializing traces and cycling the output files so that no individual file becomes to large.
- A buffer for aggregating traces to allow for batched flushes.
- An agent which initializes the tracing controller and ensures that trace serialization is done on a separate thread.
- A set of macros for generating trace events.
## Generating Trace Events

This PR introduces macros for recording trace events. Synchronous operations can be traced by wrapping them with the appropriate macros:

``` C++
TRACE_EVENT_BEGIN0("MY_CATEGORY", "SomethingCostly")
doSomethingCostly()
TRACE_EVENT_END0("MY_CATEGORY", "SomethingCostly")
```

Asynchronous operations can be traced using:

``` C++
void StartAsyncOperation(int operation_id, ...) {
  TRACE_EVENT_NESTABLE_ASYNC_BEGIN0("category", "operation", operation_id);
  ...
}
void EndAsyncOperation(int operation_id, ...) {
  TRACE_EVENT_NESTABLE_ASYNC_END0("category", "operation", operation_id);
  ...
}
```

An example of async traces can be seen below. The example was generated by integrating the above macros into Async-Wrap.

Additional examples as well as a complete list of trace macros can be found in `src/tracing/trace_event.h`. This PR does not introduce any JS interface for generating events or tracing of Node core with the hope that the community can determine how the instrumentation of core should be approached and how this functionality can be exposed to end users.
## Viewing Trace Events

The current trace serialization outputs traces in a format that can be visualized by navigating to [`chrome://tracing`](https://www.chromium.org/developers/how-tos/trace-event-profiling-tool) in Chrome and loading a trace output file.

![w85xe3osv1u](https://cloud.githubusercontent.com/assets/1485909/19749159/92196192-9b9b-11e6-803b-8c6273b12365.png)

We can zoom in on this visualization to view VM events alongside the opening and closing of TCP connections:

![dsksrjkirjkirjkk69xc63ny39g3a3zhbdtmw4aaaaaelftksuqmcc](https://cloud.githubusercontent.com/assets/1485909/19749162/964acc56-9b9b-11e6-8ea5-f5d301e2f6fa.png)
# Usage

Tracing can be enabled by running:

```
node --trace-events-enabled server.js
```

By default, this will record all events in the `"v8"` and `"node"` categories. To trace other categories or ignore either of these categories, use the `--trace-event-categories` flag:

```
node --trace-events-enabled --trace-event-categories v8,custom-category server.js
```

This PR is the result of a lot of hard work by @misterpoe, @kjin, @lpy, @fmeawad, and @ofrobots.

Thanks for the review!

/cc @nodejs/diagnostics

AndreasMadsen commented:
Awesome

> An example of async traces can be seen below. The example was generated by integrating the above macros into Async-Wrap.

Should we consider doing this by default?

matthewloring commented:
@AndreasMadsen I think that could be a good idea and it did generate useful information but I want to keep this PR focused on the core tracing system. I am hoping that instrumentation to generate trace events can be left to a follow on.

cjihrig commented:
Is it possible to add docs and tests?

matthewloring commented:
@cjihrig Working on tests now. Are the [debugger docs](https://github.com/nodejs/node/blob/master/doc/api/debugger.md) the best place to put doc changes or should they go somewhere else?

cjihrig commented:
Good question. Do you think there will be enough content to warrant a separate Tracing documentation page? If not, then the Debugger docs or even guides might be an appropriate place.

Others can weigh in here too.

sam-github commented:
In the example above, trace config doesn't look complex enough to warrant a JSON config file. It can be difficult to arrange for files on disk to appear when wanting to configure node behaviour.

I suggest the example JSON config file would be better expressed as

    node --trace-include=v8,custom_category --trace-exclude=node ...

matthewloring commented:
@cjihrig I've added an end to end test and put some documentation in the [debugger docs](https://github.com/nodejs/node/blob/master/doc/api/debugger.md). Happy to move it if there is a better place. 

@sam-github I've updated the description to list the full set of supported configuration options. Though I would anticipate included and excluded categories to be the options relevant to node, V8 does provide additional configuration.

cjihrig commented on test/parallel/test-trace-event.js:
```diff
@@ -0,0 +1,29 @@
+'use strict';
+
```
Can you include `../common`

matthewloring replied:
Done.

cjihrig commented on test/parallel/test-trace-event.js:
```diff
@@ -0,0 +1,29 @@
+const cp = require('child_process');
+
+const CODE = 'for (var i = 0; i < 100000; i++) { \"test\" + i }';
+const FILE_NAME = 'node_trace.1.log';
```
The file should probably be written to `common.tmpDir`.

matthewloring replied:
Done.

cjihrig commented on test/parallel/test-trace-event.js:
```diff
@@ -0,0 +1,29 @@
+const CODE = 'for (var i = 0; i < 100000; i++) { \"test\" + i }';
+const FILE_NAME = 'node_trace.1.log';
+
+fs.access(FILE_NAME, (err) => {
```
You can use `common.fileExists()` to avoid a level of indentation.

matthewloring replied:
Done.

cjihrig commented on test/parallel/test-trace-event.js:
```diff
@@ -0,0 +1,29 @@
+  const proc = cp.spawn(process.execPath,
+    [ '--enable-tracing', '-e', CODE ]);
+
+  proc.once('exit', () => {
```
Please add `common.mustCall()` around the callback.

matthewloring replied:
Done.

sam-github commented:
@matthewloring I still think the configuration should be via CLI options. JSON should be avoided unless necessary, where necessary means "complex data structures", which this doesn't have, its a bunch of boolean or string flags. Its not even nice for dev use... typing CLI options is easy, seeing CLI docs in the help output is easy, having to write your options into a file for a one off trace run is not easy. Sure, do it for permanent config like .eslintrc, but ephemeral config like tracing? Doesn't make sense to me.

I have tooling that allows node monitoring and tracing to be enabled in production, and I've been looking forward to the v8 trace lib getting integrated, pretty excited about this.

Restarting node with new CLI options is easy, having to write my config into filesystem (somewhere, where?) is not convenient. There isn't even a way to guarantee its cleaned up afterwards. I can do it (well, assuming I'm not running in docker with a read-only FS I can do it), but its not a good way.

Trott commented on test/parallel/test-trace-event.js:
```diff
@@ -0,0 +1,28 @@
+proc.once('exit', common.mustCall(() => {
+  assert(common.fileExists(FILE_NAME));
+  fs.readFile(FILE_NAME, (err, data) => {
+    fs.unlinkSync(FILE_NAME);
```
Tiniest of nits that can totally be ignored: No need to clean up after the test in the temp directory. Each test that needs to use the temp directory runs `common.refreshTmpDir()` which will clean up anything there.

matthewloring replied:
Fixed, thanks.

matthewloring commented:
@sam-github That makes a lot of sense. I'm happy to introduce command line flags for included/excluded categories and get rid of the json configuration. I'm going to give it a little time before changing this in case anyone has other ideas on configuration approaches.

Fishrock123 commented:
Does this have any performance impact when off?

fmeawad commented:
@Fishrock123 The performance impact is minimal specially when off

For every call site, we check if the category is enabled or disabled the first time the call site was visited, and this value is cached so that we do not have to do the expensive category check again.

If tracing was off, then it was enabled: all cached references are updated. And we do that to avoid doing the category check every time we visit the call site.

But that limits the ability to change the enabled categories when tracing is already on. Instead, we need to stop change, then start again.

That being said, trace-events should not be used to track high frequency event*, sampling and runtime stats should be used instead (both are getting standardized in tracing).
* events in tight loops.

trevnorris commented:
@fmeawad Ignoring the tight loop scenario, a single http request can make 12-18 async calls. A high speed proxy can process upwards of 30k reqs/sec. Which results in 360k-540k async calls/sec. This ignores sync calls that would also be traced (e.g. writes to a socket that are immediately flushed to the kernel). At that volume, what performance impact would we expect to see with this both enabled and disabled?

trevnorris commented:
On the side, this PR doesn't replace async hooks in any way. The two are complimentary.

jasongin commented on doc/api/debugger.md:
````diff
@@ -202,3 +202,35 @@ To start debugging, open the following URL in Chrome:
+`--trace-config` flag:
+
+```txt
+node --enable-trace --trace-config=trace-config.json server.js
````
`--enable-tracing`

matthewloring replied:
Thanks, fixed.

jasongin commented on doc/api/debugger.md:
````diff
@@ -202,3 +202,35 @@ To start debugging, open the following URL in Chrome:
+node --enable-trace --trace-config=trace-config.json server.js
+```
+
+where `trace-config.json` may override any of the default configuration values:
````
More clarification would be helpful here:
 * Are the values below actually the defaults, or just examples?
 * What are the valid values for `record_mode`?
 * What do each of those boolean flags mean?
 * Are there any other known categories for the node/v8 engine? (I realize user code can add arbitrary tracing categories.)

matthewloring replied:
Configuration has been simplified to just specify enabled categories through a command line flag. Other configuration options are not useful at this time (only one `record_mode` supported by the buffer, systrace not being used by node, etc). We are working on getting a list of categories compiled here: https://github.com/v8/v8/wiki/Tracing%20V8. Currently node does not introduce any categories of its own. The `node` category introduced by this PR is just a placeholder and can be updated when the community decides how/where trace points should be added to core.

jasongin commented on src/tracing/agent.cc:
```diff
@@ -0,0 +1,68 @@
+    std::string str((std::istreambuf_iterator<char>(fin)),
+                    std::istreambuf_iterator<char>());
+    TraceConfigParser::FillTraceConfig(env->isolate(), trace_config,
+                                       str.c_str());
```
It looks like the above code will crash if the trace config file is not found or has invalid contents. Is that acceptable? Or should there be a more helpful error message?

matthewloring replied:
Removed.

fmeawad commented:
@trevnorris According to my math, If you are using a single thread for that, an async call can take as low as 1.5us, a trace event can take up to 4us, therefore I do not recommend using trace events for that. The off overhead is going to be high as well, it will be ~10% here.

matthewloring commented:
@trevnorris Just for a data point, adding the trace events used to generate the graphs above to async-wrap caused no change in request latency and ~2% reduction in requests per second for a server with no logic:

```javascript
var server = http.createServer(function(req, res) {
  res.end('Hello world');
});
```

The % performance impact would be even lower if the server did anything before responding. We will definitely need to be careful about how many trace points are added and where they go.

jasongin commented on src/tracing/trace_config_parser.cc:
```diff
@@ -0,0 +1,116 @@
+bool TraceConfigParser::GetBoolean(Isolate* isolate, Local<Context> context,
+                                   Local<Object> object, const char* property) {
+  Local<Value> value = GetValue(isolate, context, object, property);
+  if (value->IsNumber()) {
```
Should this be:
```javascript
if (value->IsNumber() || value->IsBoolean())
```
I'm not certain `IsNumber()` is false for a boolean literal, but I'm assuming that's the case.

jasongin commented on src/tracing/node_trace_writer.cc:
```diff
@@ -0,0 +1,181 @@
+    Mutex::ScopedLock stream_scoped_lock(stream_mutex_);
+    if (total_traces_ >= kTracesPerFile) {
+      total_traces_ = 0;
+      // Destroying the member JSONTraceWriter object appends "]"\n"" to
```
`"]}"` to be more precise. (And no `\n`.)

matthewloring replied:
Fixed, thanks!

matthewloring commented on doc/api/debugger.md:
```diff
@@ -202,3 +202,35 @@ To start debugging, open the following URL in Chrome:
+  "enable_systrace": true,
+  "enable_argument_filter": true,
+  "included_categories": [ "v8", "node" ],
+  "excluded_categories": [ ]
```
This is used by Chrome since they specify enabled categories using regular expressions which can be refined by excluded categories. The node/v8 category parser doesn't support regexes so this isn't needed at this time.

matthewloring commented:
@sam-github I've updated the category configuration to use the `--enabled-categories` command line flag.

@jasongin Thank you for the review. This round of comments should all be addressed by the last commit.

trevnorris commented:
@matthewloring one area i've almost broken my back over with async hooks is that there is no overhead when not in use. the user voluntarily enabling them is a different story, but IMO a feature shouldn't impose overhead on applications that aren't using it. from what you say, it sounds like node will incur overhead for every trace event placed in core, whether they're used or not. can I get clarification that is actually the case?

matthewloring commented on doc/api/debugger.md:
```diff
@@ -202,3 +202,23 @@ To start debugging, open the following URL in Chrome:
+Node.js application.
+
+The set of categories for which traces are recorded can be specified using the
+`--enabled-categories` flag followed by a list of comma separated category names.
```
Hmm, that's a good point. Maybe `--enable-trace-events` and `--trace-event-categories`?

mhdawson replied:
Would it make sense to have them all start with --trace-event so its clear they go together.  For example:

--trace-events-enabled
--trace-event-categories

Not sure how this fits with the way we use enabled/disabled for existing options but having all options related to trace-events start with the same prefix might make it easier to know they are related.

matthewloring replied:
Updated the flags.

matthewloring commented:
@trevnorris This trace system has been designed to incur as little overhead as possible, especially when not enabled, and is already widely used in V8 and Chrome. Once categories have been cached, it should boil down to a single (easily predicted) compare and branch.

matthewloring commented:
@bnoordhuis @joshgav @mhdawson Could I get a review on this from you as well?

richardlau commented on test/parallel/test-trace-event.js:
```diff
@@ -0,0 +1,27 @@
+
+const assert = require('assert');
+const cp = require('child_process');
+const common = require('../common.js');
```
For tests, I believe `common` is supposed to be the first line (after `'use strict'`), see nodejs/node#7786.

richardlau commented:
Shouldn't the new command line options be added to the `node --help` output?

matthewloring commented:
@richardlau I've added the command line options to the `--help` output.

richardlau commented:
@matthewloring Thanks. Please could you also update `doc/api/cli.md`?

bnoordhuis commented:
@matthewloring I don't have time to do an in-depth review right now, maybe next week.

frankrousseau commented:
Hello all,
We are working on adding tracing capabilities to ExpressJS. People in this thread looks to have good knowledge of tracing problematics. So, if someone has suggestion about how to deal with it in Express, he's welcome! 
https://github.com/expressjs/express/pull/2437
PS: It's a little bit off the subject, so feel free to delete my comment if you think it should be.

Fishrock123 commented:
Could we get some numbers of what the idle and running impact of these are?

Stressing an http server with them enabled and disabled compared to current master might be ok.


Also, I'm a little concerned because from what I hear there are not necessarily at well defined points in lifecycles, so the natural reaction will just be to add them everywhere, which could both hurt performance a lot and render them far too noisy.

matthewloring commented:
@richardlau cli docs have been updated.

@Fishrock123 the idle/running impact of tracing will depend on where and how tracing is added to different subsystems of node core. For every trace point that is added, the performance impact/value tradeoff must be discussed. I am hoping to defer these conversations to the PRs that introduce the trace points themselves so that progress on the core tracing system does not become stalled.

thlorenz commented on src/node.cc:
```diff
@@ -3688,6 +3695,10 @@ static void ParseArgs(int* argc,
       trace_sync_io = true;
+    } else if (strcmp(arg, "--enable-tracing") == 0) {
+      trace_enabled = true;
+    } else if (strncmp(arg, "--enabled-categories=", 21) == 0) {
```
Not sure about requiring the `=` in `'--enabled-categories='` seems inconsistent with similar flags like `--require`

thlorenz commented on doc/api/cli.md:
```diff
@@ -129,6 +129,20 @@ added: v2.1.0
+added: v7.1.0
+-->
+
+A comma separated list of categories that should be traced when trace event
```
What kind of categories are actually available?
It's explained in the PR including how to add custom categories.
Should add that info to the docs as well or at least link to more info.

Also you gonna include some docs regarding the `config.json` file passed via `--trace-config=`? 
Like what are the keys and what they mean? I see that in the PR but nowhere as part of the docs (maybe I'm missing something).
Also again not sure if it should be `--trace-config=` vs. just `--trace-config`.

matthewloring replied:
For now the only available categories are inside V8. No special action is required to add a custom category. Adding a trace event such as `TRACE_EVENT_BEGIN0("MY_CATEGORY", "SomethingCostly")` will create the category `"MY_CATEGORY"` if it doesn't exist. As per https://github.com/nodejs/node/pull/9304#discussion_r85846385 the config file has been removed. I've updated the PR description to reflect this.

thlorenz commented on src/node.cc:
```diff
@@ -3540,6 +3544,9 @@ static void PrintHelp() {
          "  --trace-sync-io       show stack trace when use of sync IO\n"
          "                        is detected after the first tick\n"
+         "  --enable-tracing      track trace events\n"
+         "  --enabled-categories= comma separated list of trace event\n"
```
Ditto WRT `=` (as I commented in node.cc).

matthewloring replied:
Fixed.

thlorenz commented on doc/api/cli.md:
```diff
@@ -129,6 +129,20 @@ added: v2.1.0
+Enables the collection of trace event tracing information.
+
+### `--enabled-categories=list`
+<!-- YAML
```
Not sure about the `=` (see comment in node.cc).

matthewloring replied:
Fixed.

addaleax commented on doc/api/cli.md:
```diff
@@ -129,6 +129,20 @@ added: v2.1.0
 
+### `--enable-tracing`
+<!-- YAML
+added: v7.1.0
```
I’d recommend using `added: REPLACEME` instead of a fixed version here

matthewloring replied:
Fixed.

jasongin commented:
I see build errors on Windows:
```
src\tracing\agent.cc(29): error C2131: expression did not evaluate to a constant
src\tracing\agent.cc(29): note: failure was caused by non-constant arguments or reference to a non-constant symbol
src\tracing\agent.cc(29): note: see usage of 'num_chars'
src\tracing\agent.cc(31): error C3863: array type 'char [num_chars+]' is not assignable
src\tracing\node_trace_buffer.cc(132): error C4716: 'node::tracing::NodeTraceBuffer::Flush': must return a value
src\tracing\node_trace_writer.cc(62): error C2065: 'O_CREAT': undeclared identifier
src\tracing\node_trace_writer.cc(62): error C2065: 'O_WRONLY': undeclared identifier
src\tracing\node_trace_writer.cc(62): error C2065: 'O_TRUNC': undeclared identifier
src\tracing\node_trace_writer.cc(62): error C2660: 'uv_fs_open': function does not take 5 arguments
```

matthewloring commented:
@jasongin The windows build should be fixed now. However the test is currently failing with no traces being written to the output file. I can investigate but am not as familiar with libuv behavior on windows.

jasongin commented:
I debugged a little further. The underlying error is that the [call to `SetNamedPipeHandleState`](https://github.com/nodejs/node/blob/63243bcb330408d511b3945c53719425d8b7abb8/deps/uv/src/win/pipe.c#L261) returns `ERROR_INVALID_PARAMETER`. I'm not an expert on this API, but the comments in the error-handling for that call hint that it might not be possible to open a file as a pipe, at least on Windows. The [MSDN doc](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365787.aspx) implies that that function only works with named pipes, not files.

jasongin commented:
Also related feedback: should the trace writer be checking the return codes from more of the libuv calls?

trevnorris commented:
@matthewloring @thlorenz has been helping me conduct performance related analysis on this. looks like other are holding this up regardless, but would like this to hold off until i can post the numbers.

jasongin commented:
Here's a fix that works for me on Windows:
https://github.com/matthewloring/node/compare/trace-event...jasongin:tracing?expand=1
It just does an async write to the file instead of using a pipe. It also includes a bonus `std::move(str)` optimization. I haven't tested on non-Windows platforms, but there's no reason it shouldn't work fine.

matthewloring commented:
This patch is causing `test/parallel/test-trace-event.js` to fail for me on Mac. The trace output file appears to be missing a portion of the final write. Does this test pass for you on windows?

trevnorris commented:
Click on the summary below to see the diff used to generate the events for this comment.
<details>
<summary>diff used to trace `AsyncWrap` events</summary>

```diff
diff --git a/src/async-wrap-inl.h b/src/async-wrap-inl.h
index 85e31b1..7791a1c 100644
--- a/src/async-wrap-inl.h
+++ b/src/async-wrap-inl.h
@@ -13,6 +13,8 @@
 #include "util-inl.h"
 #include "v8.h"
 
+#include <string.h>
+
 namespace node {
 
 inline AsyncWrap::AsyncWrap(Environment* env,
@@ -24,6 +26,15 @@ inline AsyncWrap::AsyncWrap(Environment* env,
   CHECK_NE(provider, PROVIDER_NONE);
   CHECK_GE(object->InternalFieldCount(), 1);
 
+  switch (provider) {
+#define V(PROVIDER)                                                           \
+    case AsyncWrap::PROVIDER_ ## PROVIDER:                                    \
+      TRACE_EVENT_ASYNC_BEGIN0("async_wrap", #PROVIDER, get_uid());                \
+    break;
+    NODE_ASYNC_PROVIDER_TYPES(V)
+#undef V
+  }
+
   // Shift provider value over to prevent id collision.
   persistent().SetWrapperClassId(NODE_ASYNC_ID_OFFSET + provider);
 
@@ -68,6 +79,15 @@ inline AsyncWrap::AsyncWrap(Environment* env,
 
 
 inline AsyncWrap::~AsyncWrap() {
+  switch (provider_type()) {
+#define V(PROVIDER)                                                           \
+    case AsyncWrap::PROVIDER_ ## PROVIDER:                                    \
+      TRACE_EVENT_ASYNC_END0("async_wrap", #PROVIDER, get_uid());                  \
+    break;
+    NODE_ASYNC_PROVIDER_TYPES(V)
+#undef V
+  }
+
   if (!ran_init_callback())
     return;
 
diff --git a/src/async-wrap.cc b/src/async-wrap.cc
index 60124e4..4091102 100644
--- a/src/async-wrap.cc
+++ b/src/async-wrap.cc
@@ -26,14 +26,6 @@ using v8::Value;
 
 namespace node {
 
-static const char* const provider_names[] = {
-#define V(PROVIDER)                                                           \
-  #PROVIDER,
-  NODE_ASYNC_PROVIDER_TYPES(V)
-#undef V
-};
-
-
 class RetainedAsyncInfo: public RetainedObjectInfo {
  public:
   explicit RetainedAsyncInfo(uint16_t class_id, AsyncWrap* wrap);
@@ -191,11 +183,48 @@ void LoadAsyncWrapperInfo(Environment* env) {
 }
 
 
+class TraceMakeCallback {
+ public:
+  TraceMakeCallback(AsyncWrap::ProviderType type) : type_(type) {
+    CHECK_LT(type_, AsyncWrap::PROVIDER_TOTAL_LENGTH);
+    CHECK_GT(type_, 0);
+    switch (type_) {
+#define V(PROVIDER)                                                           \
+      case AsyncWrap::PROVIDER_ ## PROVIDER:                                  \
+        TRACE_EVENT_BEGIN1("async_wrap", #PROVIDER, "CallType", "MakeCallback");\
+      break;
+      NODE_ASYNC_PROVIDER_TYPES(V)
+#undef V
+      default:
+        fprintf(stderr, "OH CRAP!");
+        fflush(stderr);
+    }
+  }
+  ~TraceMakeCallback() {
+    switch (type_) {
+#define V(PROVIDER)                                                           \
+      case AsyncWrap::PROVIDER_ ## PROVIDER:                                  \
+        TRACE_EVENT_END1("async_wrap", #PROVIDER, "CallType", "MakeCallback");\
+      break;
+      NODE_ASYNC_PROVIDER_TYPES(V)
+#undef V
+      default:
+        fprintf(stderr, "OH CRAP!");
+        fflush(stderr);
+    }
+  }
+  AsyncWrap::ProviderType type_;
+  char* name_;
+};
+
+
 Local<Value> AsyncWrap::MakeCallback(const Local<Function> cb,
                                      int argc,
                                      Local<Value>* argv) {
   CHECK(env()->context() == env()->isolate()->GetCurrentContext());
 
+  TraceMakeCallback trace_mc(provider_type());
+
   Local<Function> pre_fn = env()->async_hooks_pre_function();
   Local<Function> post_fn = env()->async_hooks_post_function();
   Local<Value> uid = Number::New(env()->isolate(), get_uid());
@@ -210,8 +239,9 @@ Local<Value> AsyncWrap::MakeCallback(const Local<Function> cb,
     has_domain = domain_v->IsObject();
     if (has_domain) {
       domain = domain_v.As<Object>();
-      if (domain->Get(env()->disposed_string())->IsTrue())
+      if (domain->Get(env()->disposed_string())->IsTrue()) {
         return Local<Value>();
+      }
     }
   }
 
diff --git a/src/async-wrap.h b/src/async-wrap.h
index e1ea383..fe963b8 100644
--- a/src/async-wrap.h
+++ b/src/async-wrap.h
@@ -38,6 +38,14 @@ namespace node {
   V(WRITEWRAP)                                                                \
   V(ZLIB)
 
+static const char* const provider_names[] = {
+#define V(PROVIDER)                                                           \
+  #PROVIDER,
+  NODE_ASYNC_PROVIDER_TYPES(V)
+#undef V
+};
+
+
 class Environment;
 
 class AsyncWrap : public BaseObject {
@@ -47,6 +55,7 @@ class AsyncWrap : public BaseObject {
     PROVIDER_ ## PROVIDER,
     NODE_ASYNC_PROVIDER_TYPES(V)
 #undef V
+    PROVIDER_TOTAL_LENGTH,
   };
 
   inline AsyncWrap(Environment* env,
```
</details>

-----

Did some benchmarking and it looks like adding these events to the native side doesn't add any, or at least within the margin of error, overhead if the events aren't being used. And shows understandable overhead when only tracing the `'async_wrap'` events. The log output size could be interesting if the scripts are meant to run any length of time. Running the benchmark only several seconds and a few hundred MB of logs were spit out.

The diff shows that both the duration of all `MakeCallback` calls and the time an `AsyncWrap` instance is alive are being traced. I chose these because they are hit so often. Here is a screenshot of `chrome://tracing` running a benchmark with high concurrency:

![high-concurrency-http-event-tracing](https://cloud.githubusercontent.com/assets/289174/20512030/124a0b20-b03a-11e6-90a1-d39013ace75e.png)

You can see that the GUI has something to be desired when this scenario comes into play. Possibly we can help this by being more specific in our parameters?

One question left unanswered is how much overhead this will inflict if these are added from the JS side. The number of `nextTick()` calls internally is absurd, and making a call to native every time will show noticeable overhead.

In regards to performance and the native event tracers; LGTM. Though we'll have to be thoughtful in how event tracers are added throughout the code base to make them the most useful.

@matthewloring Can this be enabled using `v8.setFlagsFromString()` while the process is running?

jasongin commented:
Oh, I introduced a bug when cleaning up the code by reordering some lines: `std::move<str>` can sometimes copy instead of move the string when it's very small, that caused invalid bytes to be written. I updated the patch, still at https://github.com/matthewloring/node/compare/trace-event...jasongin:tracing?expand=1

matthewloring commented:
@jasongin The update is working for me on Mac. Thank you for the patch!

@trevnorris This currently cannot be enabled through `v8.setFlagsFromString()` to my knowledge but we could add machinery on the node side to allow for this behavior as a follow on.

matthewloring commented:
@trevnorris I just learned about a mechanism for observing when tracing has started that can react by enabling flags based on the set of enabled categories (https://cs.chromium.org/chromium/src/v8/include/v8-platform.h?sq=package:chromium&dr=C&rcl=1479817139&l=203). If the only concern is enabling tracing on an already started process, we can add a programatic api to start the trace agent at any time during execution (we will need to expose a mechanism to call https://github.com/nodejs/node/pull/9304/files#diff-cd53544f44aab2c697bcd7b6a57f23ccR4419).

matthewloring commented:
@rvagg @jbergstroem Could I get access to a smartos build machine to debug the failures there?

mhdawson commented on doc/api/debugger.md:
```diff
@@ -202,3 +202,23 @@ To start debugging, open the following URL in Chrome:
 
 [TCP-based protocol]: https://github.com/v8/v8/wiki/Debugging-Protocol
+
+## Trace Event Support for Node.js
```
Does this belong in this file ?  I see the debugger as being 2 separate things and therefore trace should be in its own file doc/api/tracing.md.

matthewloring replied:
I've moved it to doc/api/tracing.md.

mhdawson commented on doc/api/debugger.md:
```diff
@@ -202,3 +202,23 @@ To start debugging, open the following URL in Chrome:
+
+Running Node.js with tracing enabled will produce log files that can be opened
+in the [`chrome://tracing`](https://www.chromium.org/developers/how-tos/trace-event-profiling-tool)
+tab of Chrome.
```
Is there more info somewhere on this format, We should have some kind of tool available to allow the tracefile to be converted to a human readable format without requiring chrome, ideally something bundled with Node.js itself.  And then leave more sophisticated visualization/searching etc to external tools that can work with the trace file format.

matthewloring replied:
The format is currently defined by the implementation in the `JSONTraceWriter` class. The original thought was that other output formats could be supported by providing alternate implementations of `node_trace_writer.cc` to support human readable or any other desired output formats. We could also explore translating json trace output but the workflow would probably be more smooth if the output format was controlled by a flag.

mhdawson commented on src/tracing/trace_event.h:
```diff
@@ -0,0 +1,1710 @@
+
+#include <stddef.h>
+
+#include "v8-platform.h"
```
What specific parts of this file require v8-platform?  Needing this might imply that the API is tied to v8 in some way.

jasongin replied:
The `TraceEventHelper` class defined near the bottom of this file has methods that use `v8::Platform`. But actually, I found `TraceEventHelper` is not currently used anywhere: the platform is set at initialization, but never retrieved. Can it be removed? (Then replace this line with `#include <stdint.h>`.)

matthewloring replied:
The platform is set here: https://github.com/matthewloring/node/blob/51842590b093297481e995d5be23294e238d18dc/src/node.cc#L207 and used by the macros:
https://github.com/nodejs/node/blob/51842590b093297481e995d5be23294e238d18dc/src/tracing/trace_event.h#L1217
https://github.com/nodejs/node/blob/51842590b093297481e995d5be23294e238d18dc/src/tracing/trace_event.h#L1239
https://github.com/nodejs/node/blob/51842590b093297481e995d5be23294e238d18dc/src/tracing/trace_event.h#L1247

jasongin replied:
Those macros refer to `v8::internal::tracing::TraceEventHelper`, but that is different from `node::tracing::TraceEventHelper` that I was referring to.

jasongin replied:
Oh, the macro uses didn't come up in my search in VS because `src/tracing/trace_event.h` isn't referenced by `node.vcxproj` generated by the build. That header needs to be added to [`node.gyp` here](https://github.com/nodejs/node/pull/9304/files#diff-259e627be3fafed9e9c10a7fcb26879eR186).

matthewloring replied:
I've added the header.

matthewloring replied:
@mhdawson @joshgav At the diagnostics meeting today, we talked about potentially abstracting these operations on the V8 platform to a more general node platform. I do not feel strongly either way but I think it may be less churn to rewrite this once a node platform api is decided upon as opposed to starting it specifically for this use case. Either way, I wanted to continue the discussion where it could be recorded for those not at the summit.

ofrobots replied:
@mhdawson: Can this be done in a follow-on PR?

rvagg commented:
@matthewloring log in @ iojs@72.2.115.252, it's one of the smartos 16 machines. Your smartos 14 failures are compile-time but I don't think they matter because this isn't likely to be backported so it's just 15 and 16 we care about _I think_. Please let me know when you're done.

trevnorris commented:
@matthewloring it would be cool if we could enable/disable tracing, along with the categories being traced, during runtime. but it's not at all a priority.

thlorenz commented:
I agree that being able to turn tracing on/off and select categories at runtime would make this suitable for lots more use cases, especially for tracing in production.
On that note I'm not entirely clear when the log file is flushed, i.e. on my test I had to add a way to shut down the server cleanly since killing via Ctrl-C did result in the log not being flushed to disk.
(I'm on OSX sierra). Could you provide some insight here @matthewloring?
I ran into similar problems with `--perf-basic-prof` a while ago and had to find [workarounds](https://github.com/thlorenz/talks/blob/gh-pages/node-aint-just-js/slides.md#flushing-perf-pidmap).

matthewloring commented:
@rvagg I am unable to connect to this machine. Is there a process for setting up ssh keys?

matthewloring commented:
@trevnorris The tracing agent (`agent.cc`) is written so that it can be started and stopped on an already running process. We just need to figure out how we want to invoke the appropriate `Agent::Start`, `Agent::Stop` calls.

matthewloring commented:
@thlorenz The log file is flushed when `Agent::Stop` is called. This is currently only done when the process exits cleanly. Is there a path where a call to stop the agent could be added that runs when the process is killed with Ctrl-C?

joshgav commented:
I think the log file path and name should be configurable via a flag and/or config file. We don't want processes generating a lot of logs to fill up the local disk. Also, some users might want to write to a persistent or shared volume.

Should the file names include a nonce of some sort so subsequent processes in the same directory don't overwrite them?

In general, do we expect these logs to be largely temporary like dump and trace files, do we expect users to aggregate and persist them, or a mix? What has been the experience with similar logs from Chromium?

joshgav commented:
Some thoughts after a couple times through the code:

* Perhaps `trace_event.h` should be separated into two files - one with the macros intended to be utilized, and one with the typical header stuff which starts at about [line 1400](https://github.com/nodejs/node/blob/769c3b960af9a5f4b1a336569dbb30a2fd797a78/src/tracing/trace_event.h#L1399)? That would make it easier to understand I think.

* Can we use the [`traced_value.h` helper](https://github.com/v8/v8/blob/master/src/tracing/traced-value.h) in Node? That is, is it available in a public header from V8? It would help for creating structs for the `args` values.

* Only RECORDING mode is supported [in V8's TracingController currently](https://github.com/v8/v8/blob/fcf1bac99a6277234d35797847a01dfdb1f86f7a/src/libplatform/tracing/tracing-controller.cc#L104), plans and thoughts on EventCallback mode?

* Should a "Level" field be added to the static set in [trace-object](https://github.com/v8/v8/blob/master/src/libplatform/tracing/trace-object.cc)? I.e. an enum or int which reflects the severity of the entry - error, warning, info, etc. Perhaps that would be a useful index for filtering.

jasongin commented:
@joshgav Wouldn't a full implementation of a "level" field require updating all the TRACE_ macros, or at least adding new variants? Do we really want to diverge from v8 like that?

rvagg commented:
@matthewloring I used your GitHub keys, https://github.com/matthewloring.keys, pretty sure I did anyway. Give me a key here or email r@va.gg if you'd prefer and I can stick it in.

jasongin commented:
@joshgav Regarding event callbacks, maybe you saw this but there are actually [flags already defined for enabling recording and callbacks (and ETW exports) on a per-category basis](https://github.com/v8/v8/blob/master/include/libplatform/v8-tracing.h#L220). As you said, those flags are [not yet implemented](https://github.com/v8/v8/blob/master/src/libplatform/tracing/tracing-controller.cc#L135). I agree we should look into implementing the callback functionality in order to support the tracing event listener scenarios we've discussed.

thlorenz commented:
@matthewloring afaik Ctrl-C sends a `SIGINT` and I just tried to catch it properly, so this works when pressing `Ctrl-C`.

```js
process.on('SIGINT', () => console.log('caught SIGINT'))
setInterval(() => process.stdout.write('.'), 1000)
```

However there are other cases where the process just exits without a proper handler installed.
Ideally in these cases the log would be flushed as well.

ofrobots commented:
@matthewloring you might want to follow this atexit example here: https://github.com/nodejs/node/commit/ceb602326066674b894c9fc4d53166d06a918c02#diff-cd53544f44aab2c697bcd7b6a57f23ccR4478

matthewloring commented:
@thlorenz The log file should be flushed on `SIGINT` and `SIGTERM` now.

matthewloring commented:
@joshgav

- I'm definitely open to alternate file naming schemes but I would prefer holding that to a follow on PR as the current scheme is functional and there could be substantial discussion on what format makes the most sense. 

- On splitting `trace_event.h` and using the `trace_value.h` helper I think it would make sense to hold off for the separation of libtracing from V8 as this should make `trace_value.h` available and will eliminate the need for `trace_event.h`. In the mean time, it would be good to keep `trace_event.h` as close to that in V8 as possible.

- Other recording modes definitely make sense. It would be great to see more of them supported based on community needs after people have time to experiment with the system.

- Adding a separate level field may be a fairly invasive change but we could encourage including a level component in trace categories once we have more discussion on recommended category naming.

matthewloring commented:
@rvagg I am not able to connect to that machine. It looks like the key is accepted but the connection is closed. I get:
```
debug1: Authentication succeeded (publickey).
Authenticated to 72.2.115.252 ([72.2.115.252]:22).
debug1: channel 0: new [client-session]
debug3: ssh_session2_open: channel_new: 0
debug2: channel 0: send open
debug3: send packet: type 90
debug1: Requesting no-more-sessions@openssh.com
debug3: send packet: type 80
debug1: Entering interactive session.
debug1: pledge: network
debug3: send packet: type 1
debug1: channel 0: free: client-session, nchannels 1
debug3: channel 0: status: The following connections are open:
  #0 client-session (t3 r-1 i0/0 o0/0 fd 6/7 cc -1)

Connection to 72.2.115.252 closed by remote host.
Connection to 72.2.115.252 closed.
Transferred: sent 3332, received 2320 bytes, in 0.0 seconds
Bytes per second: sent 9631578.9, received 6706261.4
debug1: Exit status -1
```

matthewloring commented:
@jbergstroem Maybe you are also able to help me get access to a smartos machine?

jasongin commented on src/tracing/node_trace_writer.cc:
```diff
@@ -0,0 +1,182 @@
+  uv_async_send(&exit_signal_);
+  Mutex::ScopedLock scoped_lock(request_mutex_);
+  while(!exited_) {
+    exit_cond_.Wait(scoped_lock);
```
I think this is a windows-only issue, since libuv's unix implementation of async handles does not have the same concept of endgame processing.

jasongin replied:
The best solution I can come up with is to add a `Close()` method to `NodeTraceWriter` and `NodeTraceBuffer`, so that the async handles can be closed without deleting those objects. Then they can be safely deleted after the loop thread exits. I'm working on a patch.

jasongin replied:
Fixed in my fork here:
https://github.com/matthewloring/node/compare/trace-event...jasongin:tracing?expand=1

matthewloring replied:
I've pulled in the patch. Thank you!

jasongin commented:
I found that the v8 implementation of the `TraceConfig` class does not support emitting **multi-category events**, despite the [comments in `trace_event_common.h`](https://github.com/nodejs/node/blob/master/deps/v8/base/trace_event/common/trace_event_common.h#L30) (and many other places) indicating that should be supported. I think the only deficiency is in the implementation of [`TraceConfig::IsCategoryGroupEnabled()`](https://github.com/nodejs/node/blob/master/deps/v8/src/libplatform/tracing/trace-config.cc#L23), where it incorrectly assumes the `category_group` argument is a single category; instead it needs to split a potentially comma-separated list of categories and check each one separately. (The result gets cached, with the group as the key, so doing that extra work would not be a significant performance concern.)

Unforutunately the `TraceConfig` class is not designed to be overridable, so I think this would need to be fixed in v8 sources.

So how much do we care about multi-category events? Probably we could get by without it when just instrumenting the node engine. But if we intend to expose tracing APIs to user code then I think that's a fairly important feature.

jasongin commented:
I patched my fork here (https://github.com/jasongin/nodejs/commit/dcc50445a364664586164f42f9250732bd372982) and now multi-category events are working as expected. But I don't know that we want to diverge from the v8 copy of the code.

matthewloring commented:
@joshgav I've added the experimental warning discussed at the summit.

fmeawad commented:
@jasongin We wanted to keep the V8 implementation as simple as possible and then come back to these issues when the need arise. If there is a need for multiple categories, then we should add it to V8. If you have a patch, please send it to me.

matthewloring commented:
@fmeawad Patch can be found here https://github.com/jasongin/nodejs/commit/dcc50445a364664586164f42f9250732bd372982.

@jasongin `TraceConfig` should be accessible to be overridden. A change like this probably makes more sense to land in V8 so that all trace users can benefit but I'm curious what was meant by  the TraceConfig class is not designed to be overridable".

lpy commented:
@jasongin I will do a patch in V8 to address this issue.

jasongin commented:
@lpy That would be great, thanks!

@matthewloring by "not designed to be overridable" I meant none of the `TraceConfig` methods are virtual. So a subclass would not be able to change any behavior of existing methods, which would be necessary to add multi-category support.

matthewloring commented:
@jasongin Ah right, makes sense. Change has been opened here: https://codereview.chromium.org/2558193002/.

jasongin commented:
@joshgav wrote earlier:
> I think the log file path and name should be configurable via a flag and/or config file. We don't want processes generating a lot of logs to fill up the local disk. Also, some users might want to write to a persistent or shared volume. Should the file names include a nonce of some sort so subsequent processes in the same directory don't overwrite them?

I share these concerns. As I've been experimenting with tracing, I've seen how limiting it is for the log file to be always the same filename in the current directory. I think the default log filename should be unique, and the directory should be settable via a command-line option and/or env variable.

matthewloring commented:
@jasongin I completely agree. I am just hoping to leave that change to a follow on PR along with some of the other enhancements that have been discussed recently. I want to land this as soon as I can get access to a smartos machine to debug the failures there.

/ping @rvagg @jbergstroem

joshgav commented:
The other remaining open item from [the Node Interactive discussion](https://github.com/nodejs/diagnostics/blob/master/wg-meetings/2016-12-02.md#next-steps-for-tracingcontroller) is whether the TracingController in this PR should be further abstracted from V8 now.

From https://github.com/nodejs/node/pull/9304#discussion_r90739052 above:

@matthewloring 

> abstracting these operations on the V8 platform to a more general node platform

@mhdawson

> we should make it general now

What I was referring to at Node Interactive is that the V8 platform, at least in the literal sense of `v8::v8_platform`, is now somewhat abstracted in Node through this construct: [src/node.cc#L200](https://github.com/nodejs/node/blob/df3978421b86991fb45256f1a8694e0590019b13/src/node.cc#L200)

This PR (#9304) puts the TracingController support in the same abstraction, see [src/node.cc#L209](https://github.com/matthewloring/node/blob/4adececbfcf36481003a5df2c0b668a502e36046/src/node.cc#L209).

So I think abstracting out that struct as `node::platform` or somesuch, with `v8::v8_platform` deriving from it, is likely the proper solution both for TracingController and other platform-level abstractions. Question is should that abstraction hold up this PR, or should we pursue it separately? I think the latter, and perhaps we should start that pursuit in another issue now.

ofrobots commented:
@jbergstroem @rvagg @nodejs/build can we make it easier to get access to build environments specially for platforms like SmartOS which collaborators are unlikely to get access to. It seems that @mattloring has been asking for access for a while without success.

bnoordhuis commented:
The smartos14 build errors seem to be because of antiquated STL headers.  I think we're about to drop support for smartos14; it should be safe to ignore.

Since smartos is not a tier 1 platform I think it's okay to simply mark the failing test as flaky on smartos15.  @nodejs/platform-smartos can take a look if they want.

The freebsd failure looks unrelated, probably a flake.

jbergstroem commented:
@ofrobots: i've been a bit overloaded; reaching out to him now. the best procedure is always to open an issue at the build repo so we get more eyes on it.

joshgav commented:
@mhdawson can you respond to https://github.com/nodejs/node/pull/9304#issuecomment-265822544?

@matthewloring @jasongin I'd feel more comfortable if the log path was configurable now (to avoid issues like [this](https://gist.github.com/joshenders/b683c753fe37df9d76dc1e9a0157dff1) or filling up the local directory), but if everyone else feel it's okay to land in master as is then I'll relent :).

matthewloring commented:
It looks like smartos is failing on calls to `base::ThreadTicks::Now()` on trace event initialization. Specifically, we are hitting the `UNREACHABLE` case [here](https://cs.chromium.org/chromium/src/v8/src/base/platform/time.cc?q=base::ThreadTicks::Now&sq=package:chromium&l=652). I've opened a V8 issue [here](https://bugs.chromium.org/p/v8/issues/detail?id=5739).

joshgav commented:
Conclusion in today's CTC meeting is to defer work on abstraction from V8 till after this lands, since we're landing in master and clearly marked experimental.

With that, LGTM!

jasongin commented:
LGTM

I wrote some notes about how I'm building on this work [in the related diagnostics WG issue](https://github.com/nodejs/diagnostics/issues/53#issuecomment-267193097).

matthewloring commented:
I believe I have all the necessary LGTMs on this. I will land it as soon as https://github.com/nodejs/node/pull/10342 goes through to fix the smartos issues.

matthewloring commented:
It would be good to squash all these commits at landing time but I don't want to lose attribution for different people's contributions. Is there a protocol for attributing multiple people to a single commit?

matthewloring commented:
@Trott ^

addaleax commented:
> Is there a protocol for attributing multiple people to a single commit?

I don’t think so. You can try to squash this into logical units that each pass testing individually and have proper attribution, but I get that that may not be feasible; maybe just invent a new `Author: Foo Bar <foobar@gmail.com>` metadata field and add corresponding lines at the bottom of the commit message.

matthewloring commented:
@addaleax I explored the first option and it wasn't looking promising. @jasongin How do you feel about the author tag approach?

matthewloring commented:
Final CI run is green: https://ci.nodejs.org/job/node-test-commit/6746/.

The arm failure showing up on the PR looks like a case of https://github.com/nodejs/github-bot/issues/108.

jasongin commented:
@matthewloring That's fine with me, I don't care much about having my name on the commit.

Qard commented:
It's really just the code ownership attribution that matters. We want to be able to find the authors if a problem comes up in the future. As long as there's *some* metadata in there connecting the relevant people to the commit, it should be all good I think.

matthewloring commented:
Landed in: ba4847e879424ad173289e8fb96cc86a09ee899b. Thank you for the reviews and contributions!

targos commented:
Is there a plan to backport this to release branches?

evanlucas commented:
@matthewloring This isn't landing cleanly on v7.x. Mind submitting a backport PR?

richardlau commented:
If this is backported please also include #10959.

matthewloring commented:
@evanlucas I opened a backport PR here: https://github.com/nodejs/node/pull/11106.

Trott commented:
Adding possibly-gratuitous but hopefully-not-harmful `do-not-land` labels for 4.x and 6.x. /cc @MylesBorins (3 of 3)

jasnell commented:
ping @nodejs/lts ... do we want this in 4 and 6?
/cc @nodejs/ctc ... thoughts?

matthewloring commented:
The versions of V8 in 4 and 6 do not have the APIs that this change depends on.
